clean-modular-code/python-pep-8.md

```diff
@@ -224,27 +228,103 @@ plt.show()
+
+Ruff is both a linter and code formatter which is why we like to use it when developing both software and .py file based workflows. 
+:::
 
 +++ {"editable": true, "slideshow": {"slide_type": ""}}
 
-## Additional Resources
+DRAFT!! 
+
+
+## Best Practices for Importing Python Packages In Scientific Code
+
+There are a set of best practices that you should follow when importing **Python** packages in your code. These best practices are outlined in the <a href="https://www.python.org/dev/peps/pep-0008/#imports" target="_blank">PEP 8 guidelines</a> and apply to both **Python** scripts and to working in **Jupyter Notebook** files.
+
+## Import Python Libraries at the Top of Your Script or Notebook
+
+It is good practice to import all of the packages that you will need at the top of your **Python** script (.py file) or in the first code cell of a **Jupyter Notebook** file. 
+
+This allows anyone looking at your code to immediately know what packages they need to have installed in order to successfully run the code. This rule also follows the PEP 8 conventions for **Python** code.
+
+```python
+import os 
+
+import pandas as pd
+import numpy as np
+```
+
+Once you have imported all of the packages that you need to run your code in a script, you have access to all of the functions and classes defined in each package. 
+
+If these imports are at the top of the script or **Jupyter Notebook** file, then you will be able to use those packages in any code lines that follow. 
+
+This means that if you import a package *after* running some code that requires that package, your code will not run successfully.
+
+
+## 2. List Package Imports Following PEP 8 Standards: Most Common First, Followed By Third Party
+
+<a href="https://www.python.org/dev/peps/pep-0008/#imports" target="_blank">PEP 8</a> also specifies the order in which you should list your imports as follows (starting with the most commonly used):
+
+> Imports should be grouped in the following order:
+>    Standard library imports.
+>    Related third party imports.
+>    Local application/library specific imports.
+> You should put a blank line between each group of imports.
+
+You may be wondering, what is a standard library import? The standard library imports are commonly used tools that are general in purpose and are part of the standard library of **Python**. These including things like:
+
+* **os**: handle files and directories.
+* **glob**: create lists of files and directories for batch processing.
+
+In the PEP 8 order, other commonly used packages that are general in purpose will follow such as: 
+
+* **numpy**: work with data in array formats.
+* **matplotlib**: plot data.
+* * **Pandas**: to work with tabular data.
+
+A PEP 8 order of imports for commonly used **Python** packages for science would look something like this:
+
+```python
+import os
+import glob
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
```

Comment:
add the third block of 

```python
# imports from your local package
import mypackage
from mypackage.internal import _secret_function
```

Reply:
oh i see you have a separate section for that. my b. i just figured showing all three of them in the example that follows the list of 3 would be nice, but this is a draft and i'm jumping the gun on giving comments lol

Reply:
go for it!! comment away. i just combined two pages. i'm totally fine with consolidating!!  

 This is good, too, because in our second workshop, they will create a package and import it. so they will get to try this out!!

Reply:
I've just pushed a new version of this and made all of the fixes that you suggested. I also redesigned the structure to be:

* Example of good vs not as good code format. 
* How to apply tools to help you make your code prettier :) 
* Why use these tools and standards
* Some rules to remember 

That way, we start with something that doesn't have to be so hard and end with the details that many people will skip (I think?).

Reply:
I hear you on that but i think that currently it's a little bewildering the other way where we're missing some introductory information on what we're talking about and why first, about to post the review

edit: oh wait it makes this comment part of the review even when we started it before that lol my b

Reply:
😆 no worries!! i think / hope that i addressed your comment!! I am going to focus on merging this today give we are a week out BUT we can always revisit all of this content and update/ enhance / beautify, etc in the future!! 🚀 

thank you for this feedback!!